### ISSUE TITLE
How to get the complete folder path on click of the any child or sub folder directly 

### ISSUE THREAD

**Issue by vinaymuthanna0727** [GleamTech/FileUltimate#5]
Hi ,
I need to get the complete folder path on click of the child folder directly 
![fileultimate](https://user-images.githubusercontent.com/10174962/44249958-f301f100-a20f-11e8-87e6-39f8de5f1986.PNG)

Attached the image .
After opening multiple folder and sub folder in file ultimate and i will directly click on any one of the sub folder .Now i need to the get the complete path of the parent folder 

In Image When i click on Metadata sub folder directly .Need to get complete path including the parent folder details .
Any there any solution to get the complete file path on clicking the child or sub folder directly in file ultimate ??

**Comment by GleamTech** [comment]
There is no direct event for it (we plan to add it in future) but for now you can use this code:
```html
<head> 
  <script type="text/javascript">
  
    //Override for adding getNavigationSelectionInfo amd getViewSelectionInfo methods
    //and fileManagerNavigationSelectionChange and fileManagerViewSelectionChange events
    function fileManagerLoading(sender, e) {
        var fileManager = sender;

        Ext.override(fileManager, {
            getNavigationSelectionInfo: function () {
                var record = this.navigationSelection;

                if (!record)
                    return null;

                return {
                    Name: record.data.name,
                    FullPath: record.getPathData().fullPath,
                    ItemType: Ext.Object.getKey(GleamTech.FileUltimate.FileManagerItemType, record.data.itemType)
                };
            },
            getViewSelectionInfo: function() {
                var parentFullPath = this.navigationSelection.getPathData().fullPath;

                var items = [];
                for (var i = 0; i < this.viewSelection.length; i++) {
                    var record = this.viewSelection[i];

                    items.push({
                        Name: record.data.name,
                        FullPath: GleamTech.Util.Path.combine(parentFullPath, record.data.name, GleamTech.Util.Path.backSlash),
                        ItemType: Ext.Object.getKey(GleamTech.FileUltimate.FileManagerItemType, record.data.itemType),
                        Extension: record.data.extension,
                        Size: record.data.size
                    });
                }

                return items;
            },
            onNavigationSelectionChange: function() {
                this.callParent(arguments);

                var eventArgs = this.getNavigationSelectionInfo();

                if (eventArgs)
                    Ext.callback(window["fileManagerNavigationSelectionChange"], window, [this, eventArgs]);
            },
            onViewSelectionChange: function () {
                this.callParent(arguments);

                var eventArgs = this.getViewSelectionInfo();

                Ext.callback(window["fileManagerViewSelectionChange"], window, [this, eventArgs]);
            }
        });
    }

  </script>
</head> 
<body>

  <GleamTech:FileManager id="fileManager" runat="server" 
    ClientLoading="fileManagerLoading" />
  
</body>
```

Then for navigation selection:
```js
//OPTION 1:
function fileManagerNavigationSelectionChange(sender, e) {
    //Pretty print the navigation selection (from event object) when it changes
    alert(JSON.stringify(e, null, 2));
}

//OPTION 2:
function myButton1Click() {
    //Pretty print the navigation selection when the your button is clicked
    var selection = fileManager.getViewSelectionInfo();
    alert(JSON.stringify(selection, null, 2));
}

//SAMPLE OUTPUT:
/*

{
  "Name": "Folder One",
  "FullPath": "[1. Root Folder]:\\Folder One",
  "ItemType": "Folder"
}

*/
```

Then for view selection:
```js
//OPTION 1:
function fileManagerViewSelectionChange(sender, e) {
    //Pretty print the view selection (from event object) when it changes
    alert(JSON.stringify(e, null, 2));
}

//OPTION 2:
function myButton2Click() {
    //Pretty print the view selection when the your button is clicked
    var selection = fileManager.getNavigationSelectionInfo();
    alert(JSON.stringify(selection, null, 2));
}


//SAMPLE OUTPUT:
/*

[
  {
    "Name": "DOCX Document.docx",
    "FullPath": "[1. Root Folder]:\\DOCX Document.docx",
    "ItemType": "File",
    "Extension": "docx",
    "Size": 84272
  },
  {
    "Name": "Folder One",
    "FullPath": "[1. Root Folder]:\\Folder One",
    "ItemType": "Folder",
    "Extension": "",
    "Size": null
  }
]

*/
```

**Comment by GleamTech** [comment]
FYI, we have just released FileUltimate v6.2.0 with this featue:
 
- Added: New client-side events. Before events: ClientCreating, ClientDeleting, ClientRenaming, ClientCopying, ClientMoving, ClientCompressing, ClientExtracting, ClientUploading, ClientDownloading, ClientPreviewing. It's possible to stop a file action (and optionally display a message) by canceling the corresponding before event. After events: ClientFolderChanged, ClientSelectionChanged, ClientCreated, ClientDeleted, ClientRenamed, ClientCopied, ClientMoved, ClientCompressed, ClientExtracted, ClientUploaded. Refer to new "Client-side events" sample and updated docs for event handler usage.
 
Now you don't need any overrides anymore and you can directly use these client side events:
```html
​<head> 
  <script type="text/javascript">
  
    function fileManagerFolderChanged(sender, e) {
        var fileManager = sender;

        //Pretty print the event arguments
        var json = JSON.stringify(e, null, 2);

        alert(json);
    }

    function fileManagerSelectionChanged(sender, e) {
        var fileManager = sender;

        //Pretty print the event arguments
        var json = JSON.stringify(e, null, 2);

        alert(json);
    }

  </script>
</head> 
 ```

Binding JS event handlers in ASP.NET MVC:
```c#
var fileManager = new FileManager
{
    ClientFolderChanged="fileManagerFolderChanged",
    ClientSelectionChanged="fileManagerSelectionChanged"
};
```

Binding JS event handlers in ASP.NET WebForms:
```aspx
<GleamTech:FileManager id="fileManager" runat="server" 
  ClientFolderChanged="fileManagerFolderChanged"
  ClientSelectionChanged="fileManagerSelectionChanged" />
```

Sample outputs for these event handlers (pretty printing of event arguments):
```js
​//SAMPLE OUTPUT for ClientFolderChanged:
{
  "eventName": "folderChanged",
  "folder": {
    "name": "Folder One",
    "fullPath": "[1. Root Folder]:\\Folder One",
    "itemType": "Folder"
  }
}

//SAMPLE OUTPUT for ClientSelectionChanged:
{
  "eventName": "selectionChanged",
  "items": [
    {
      "name": "DOCX Document.docx",
      "fullPath": "[1. Root Folder]:\\DOCX Document.docx",
      "itemType": "File",
      "extension": "docx",
      "size": 84272
    },
    {
      "name": "Folder One",
      "fullPath": "[1. Root Folder]:\\Folder One",
      "itemType": "Folder",
      "extension": "",
      "size": null
    }
  ]
}
```

Getting current folder info and selection info when your buttons are clicked:
```js
//Pretty print the current folder info when the your button is clicked
function myButton1Click() {
    var folderInfo = fileManager.getFolderInfo();
    var json = JSON.stringify(folderInfo, null, 2);

   alert(json);
}

//SAMPLE OUTPUT:
/*
{
  "name": "Folder One",
  "fullPath": "[1. Root Folder]:\\Folder One",
  "itemType": "Folder"
}
*/

//Pretty print the selection info when the your button is clicked
function myButton2Click() {
    var selectionInfo = fileManager.getSelectionInfo();
    var json = JSON.stringify(selectionInfo, null, 2);

   alert(json);
}

//SAMPLE OUTPUT:
/*
[
  {
    "name": "DOCX Document.docx",
    "fullPath": "[1. Root Folder]:\\DOCX Document.docx",
    "itemType": "File",
    "extension": "docx",
    "size": 84272
  },
  {
    "name": "Folder One",
    "fullPath": "[1. Root Folder]:\\Folder One",
    "itemType": "Folder",
    "extension": "",
    "size": null
  }
]
*/
```

**Comment by NiteshSaxena** [comment]
One easy way to do this as a workaround which works well is to change the "ListFolder" and "ExpandFolder" code in "FileManagerHandler.cs".

You can call List Folder recursively and exclude a special folder by matching with its name, and get data from the recursive call and add to the result set which is sent to the browser, This way you can skip the special directories.

```
foreach (var fileSystemEntry in fileSystemEntries)
{
     ....
     if (fileSystemEntry.IsFolder)
     {
           if (fileSystemEntry.Name.ToLower().Contains("metadata"))
           {
                var collection = ListFolder(Path.Combine(path, fileSystemEntry.Name), false);
               if (collection == null || ((collection.Files == null || collection.Files.Count == 0) && (collection.Folders == null || collection.Folders.Count == 0)))
                     continue;

               if (collection.Files.Count > 0)
                     result.Files.AddRange(collection.Files);

               if (collection.Folders.Count > 0)
                     result.Files.AddRange(collection.Folders);

               continue; //To keep looping.
          }
         ....
     }
     ....
}
```
